pandas/_libs/interval.pyi

```diff
@@ -0,0 +1,96 @@
+from datetime import timedelta
+import numbers
+from typing import (
+    Generic,
+    TypeVar,
+)
+
+import numpy as np
+import numpy.typing as npt
+
+from pandas._typing import (
+    IntervalCloseness,
+    Scalar,
+)
+
+_ScalarT = TypeVar("_ScalarT", bound=Scalar)
+
+class IntervalMixin:
+    @property
+    def closed_left(self) -> bool: ...
+    @property
+    def closed_right(self) -> bool: ...
+    @property
+    def open_left(self) -> bool: ...
+    @property
+    def open_right(self) -> bool: ...
+    @property
+    def mid(self) -> float: ...
+    @property
+    def length(self) -> float: ...
+    @property
+    def is_empty(self) -> bool: ...
+    def _check_closed_matches(self, other: IntervalMixin, name: str = ...) -> None: ...
+
+class Interval(IntervalMixin, Generic[_ScalarT]):
+    def __init__(
+        self,
+        left: _ScalarT,
+        right: _ScalarT,
+        closed: IntervalCloseness = ...,
+    ) -> None: ...
+    @property
+    def closed(self) -> str: ...
+    @property
+    def left(self) -> _ScalarT: ...
+    @property
+    def right(self) -> _ScalarT: ...
+    def __str__(self) -> str: ...
+    # TODO: could return Interval with different type
+    def __add__(
+        self, y: numbers.Number | np.timedelta64 | timedelta
+    ) -> Interval[_ScalarT]: ...
+    def __radd__(
+        self, y: numbers.Number | np.timedelta64 | timedelta
+    ) -> Interval[_ScalarT]: ...
+    def __sub__(
+        self, y: numbers.Number | np.timedelta64 | timedelta
+    ) -> Interval[_ScalarT]: ...
+    def __mul__(self, y: numbers.Number) -> Interval[_ScalarT]: ...
+    def __rmul__(self, y: numbers.Number) -> Interval[_ScalarT]: ...
+    def __trudeiv__(self, y: numbers.Number) -> Interval[_ScalarT]: ...
+    def __floordiv__(self, y: numbers.Number) -> Interval[_ScalarT]: ...
+    def overlaps(self, other: Interval[_ScalarT]) -> bool: ...
+
+VALID_CLOSED: frozenset[str]
+
+def intervals_to_interval_bounds(
+    intervals: np.ndarray, validate_closed: bool = ...
```

Comment:
do we know anything about intervals.dtype?

Reply:
Yes, I added comments above. Unfortunately, we cannot have npt.NDArray[Interval] since Interval does not inherit from np-generic.